app/models/course/experience_points_record.rb

```diff
@@ -30,4 +32,34 @@ def manually_awarded?
   def reason
     manually_awarded? ? super : specific.experience_points_display_reason
   end
+
+  private
+
+  def send_notification
+    Course::LevelNotifier.level_reached(course_user.user, level_after_update)
+  end
+
+  # Test if the course_user will reach a new level after current update.
+  def reached_new_level?
+    return false unless points_awarded && points_awarded_changed?
+
+    current_level = level_before_update
+    new_level = level_after_update
+
+    current_level.persisted? &&
+      new_level.persisted? &&
+      new_level.level_number > current_level.level_number
+  end
+
+  def level_before_update
+    current_exp = course_user.experience_points
+    course_user.course.level_for(current_exp)
+  end
+
+  def level_after_update
+    # Since we are in the before_save callback, exp changes are not saved yet.
+    exp_changed = points_awarded - (points_awarded_was || 0)
+    current_exp = course_user.experience_points
```

Comment:
@lowjoel `experience_points` is a calculated attributes, so it requires `course_user` to be persisted.

But in our experience_point_records factory, the course_user is not persisted during creation (https://github.com/Coursemology/coursemology2/blob/master/spec/factories/course_experience_points_records.rb#L11), discussed with @weiqingtoh and we have two options:

1. Change the experience_point_records factory and make the `course_user` persisted before it was created.
2.  Not using the method provided by calculated_attributes and manually sum the exp : ` course_user.experience_points_records.pluck(:points_awarded).sum`
3. Redefine the instance method `experience_points` in course_user, make it work even when course_user is not persisted. 

Option 2 breaks the abstraction and duplicates the logic in course_user.
Both option 2 and 3 feels like we are modifying the actual code to suit the test ( or factory_girl ), because non-perissited course_user only happens during the creation of the factory. 

After discussion, we lean towards option 1, since the test suite should follow actual usage of the app, rather than the other way around. 

However this breaks the `build_stubbed` thing for certain tests, so we have to converted it into `create` instead. one example is: https://github.com/Coursemology/coursemology2/blob/master/spec/models/course/assessment/question/multiple_response_spec.rb#L25

Are you OK with this ? 

Thanks !